### ISSUE TITLE
what does it mean?

### ISSUE THREAD

**Issue by Esther-qian** [DominikBatic/EndoViT#2]
In the end, we haven't successfully integrated ViT into RDV. Therefore, we test using a simple model consisting of a feature extraction backbone (either ResNet50 or ViT) and a single linear layer.

**Comment by DominikBatic** [comment]
Hello,

After pretraining EndoViT, we wanted to test out how it works as a feature extraction backbone in various downstream tasks. Therefore, we looked for models that already have a feature extraction backbone we could replace with EndoViT.

After some research we decided to go with the following tasks / models:
1) Action Triplet Detection
- Rendezvous: [https://github.com/CAMMA-public/rendezvous?tab=readme-ov-file](https://github.com/CAMMA-public/rendezvous?tab=readme-ov-file)
2) Surgical Phase Recognition
- TeCNO: [https://github.com/tobiascz/TeCNO/tree/master?tab=readme-ov-file](https://github.com/tobiascz/TeCNO/tree/master?tab=readme-ov-file)
3) Semantic Segmentation
- Here we train our own model using Dense Prediction Transformer: [https://github.com/isl-org/DPT](https://github.com/isl-org/DPT)

---

The first step was always to reproduce the results of the original paper. However, with Rendezvous we were stuck for a month. At that time they didn't release any of the pretrained weights, so we trained the model from scratch ourselves. We were getting lower metric values on the test set than what they report. And to make it more confusing, the results were quite different between the runs.

Then we noticed in their GitHub issues that other people were having similar problems.

---

We analyzed Rendezvous code and discovered some implementation issues in the Pytorch version. We raised an issue in the GitHub repo and got a confirmation from the author that it was indeed a mistake:

- https://github.com/CAMMA-public/rendezvous/issues/10#issue-1518742980

Further analysis of the code, showed some inconsistencies between what is written in the paper and the implementation itself, 
- e.g. Attention Is All You Need paper (https://arxiv.org/abs/1706.03762) introduces attention as:

<p align="center">
  <img src="https://github.com/DominikBatic/EndoViT/assets/66787240/3b72338f-1c51-4306-bb28-31f2b33efc8f" alt="attn_org" width="75%">
</p>

In their paper they use a slightly different, but equally correct formula:

<p align="center">
  <img src="https://github.com/DominikBatic/EndoViT/assets/66787240/6cd50f25-9e00-4be5-836d-f00934ada1f6" alt="attn_rdv" width="60%">
</p>

But if you look at "network.py" file in the Rendezvous repository. In "MHMA" module you will find the following implementation of the attention mechanism:

https://github.com/CAMMA-public/rendezvous/blob/45181766ced1b154edf4ec27006b1d050d797de0/pytorch/network.py#L329-L335

It follows the formula in their paper, until the last matrix multiplication where they multiple A * V instead of V * A, which in our opinion is not correct.

There are other inconsistencies, like vector dimensions being transposed. However, it is hard to explain all the details.

---

To sum up, in our opinion both modules that contain attention, i.e. CAGAM and MHMA are not correctly implemented in the Pytorch implementation. Because we were not sure if there are any other mistakes in the code, we decided we didn't have enough time to fully integrate MAE into Rendezvous model and proceeded to implement the other 2 downstream tasks.

Here is the architecture of Rendezvous. As you can see it has a feature extractor ResNet18 backbone, followed by an encoder and a decoder module.

<p align="center">
  <img src="https://github.com/DominikBatic/EndoViT/assets/66787240/2f4efd92-dffa-4a18-aac0-425071bc62e5" alt="rdv_architecture" width="50%">
</p>

Instead of the original plan to use Rendezvous and replace its feature extractor with EndoViT, we build a very simple model. We replace the Rendezvous' encoder/decoder with 1/2 linear layers and use ResNet18 or EndoViT as a backbone. The results we reporeted were obtained with this architecture.

---

P.S. EndoViT is already integrated into Rendezvous as a backbone in our repository, just take a look at the "network.py" file:
- https://github.com/DominikBatic/EndoViT/blob/main/finetuning/action_triplet_detection/model/Rendezvous/network.py

I went through the code of Rendezvous again, and I believe this should be the correct implementation. These are the changes that I made:
-  in MHMA module (Rendezvous code left vs Ours right):

<p align="center">
  <img src="https://github.com/DominikBatic/EndoViT/assets/66787240/bc261768-8ac9-4239-8e05-f4942ee68157" width="100%" alt="MHMA_changes">
</p>

- in CAGAM module:

<p align="center">
  <img width="100%" alt="CAGAM_changes" src="https://github.com/DominikBatic/EndoViT/assets/66787240/c2d71158-2d0a-4cde-b9a0-a55e0e3ef0ff">
</p>

---

Keep in mind that we never tested if the code works correctly. If you want to run it yourself, feel free to and I will help you as much as I can.

One thing that you should pay attention to is the following: Rendezvous extracts High and Low level features out of the ResNet backbone. It remains to be decided after which layer of EndoViT you should extract the features, since they have completely different architectures. We did so after 4th and 12th Attention block, as you can see here:

https://github.com/DominikBatic/EndoViT/blob/1c87047a39cb2fe83db28916d0748af30542e8e7/finetuning/action_triplet_detection/model/Rendezvous/network.py#L162-L163

However, we never confirmed if this is the best choice with a concrete analysis.

Hope this explained your question. If you have any further remarks, feel free to ask.
